crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -4508,20 +4509,19 @@ impl<'a> Checker<'a> {
     where
         'b: 'a,
     {
-        let mut stacks = vec![];
+        let mut stacks = Vec::with_capacity(self.deferred.string_type_definitions.len());
         self.deferred.string_type_definitions.reverse();
-        while let Some((range, expression, (in_annotation, in_type_checking_block), deferral)) =
+        while let Some((range, value, (in_annotation, in_type_checking_block), deferral)) =
             self.deferred.string_type_definitions.pop()
```

Comment:
Nit: We can try to use `drain` to avoid the need for reversing and manually call pop

```rust
let mut string_type_definitions = self.deferred.string_type_definitions.drain(..).rev();
for ((range, value, (in_annotation, in_type_checking_block), deferral)) = string_type_definitions {
	...
}
```

If the borrow checker gets mad because of multiple mutable borrows of `self`, then try this:

```rust
let mut string_type_definitions = std::mem::take(&mut self.deferred.string_type_definitions);
for ((range, ..)) = string_type_definitions.drain(..).rev() {
  ..
}

// Restore vector, to re-use the heap allocations
self.deferred.string_type_definitions = string_type_definitions; 
```

Both these approaches assume that the loop's body does not read from or write to `self.deferred.string_type_definitions.

Reply:
I think we actually _do_ want to write to `self.deferred.string_type_definitions`, but we probably _don't_ do it currently (https://github.com/charliermarsh/ruff/issues/3655).